### DISCUSSION TITLE
Server Error (500) - gunicorn

### DISCUSSION THREAD

**Discussion by pmatuszy** [healthchecks/healthchecks#740] [Q&A]
Hi
I am trying to fix my installation.
I managed to start the webapp running so in the logfile I can see responses to pings
```
[2022-11-24 12:03:47 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:49 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:51 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:53 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:55 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:56 +0000] [67790] [DEBUG] POST /ping/6cb77b0e-0006-4ef8-94d9-0ba19dd3e972
[2022-11-24 12:03:56 +0000] [67790] [DEBUG] Closing connection.
[2022-11-24 12:03:57 +0000] [67790] [DEBUG] POST /ping/7873c9ae-e215-464c-ad7c-f7685cb18b57
[2022-11-24 12:03:57 +0000] [67790] [DEBUG] Closing connection.
[2022-11-24 12:03:57 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:03:59 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:01 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:03 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:05 +0000] [67790] [DEBUG] GET /ping/4db57544-ba71-40f9-b9e3-6a179e5d4871
[2022-11-24 12:04:05 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:05 +0000] [67790] [DEBUG] Closing connection.
[2022-11-24 12:04:06 +0000] [67790] [DEBUG] POST /ping/6cb77b0e-0006-4ef8-94d9-0ba19dd3e972
[2022-11-24 12:04:06 +0000] [67790] [DEBUG] Closing connection.
[2022-11-24 12:04:07 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:09 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:11 +0000] [67790] [DEBUG] GET /
[2022-11-24 12:04:13 +0000] [67790] [DEBUG] GET /

```
so far so good. The problem is with web interface - when I open it I get "Server Error (500)".
I have no clue how to debug it - in logfiles (errorlog and accesslog) there is nothing about it.
Could you point me to any place where I can find the reason for website misbehaviour?
FYI: I used to receive emails from my site so EMAIL_* settings are good but now I receive no errors via email, so I am a bit blind...

Regards
Paul

**Comment by cuu508** [comment]
Are you running with `DEBUG=False`?

**Comment by pmatuszy** [comment]
affirm - DEBUG is set to False 

"export DEBUG=False"

**Comment by cuu508** [comment]
Have you run `manage.py compress`, `manage.py collectstatic` and `manage.py migrate`?

https://github.com/healthchecks/healthchecks#running-in-production

**Reply by pmatuszy** [reply]
yes - all was run:
```
./manage.py compress
CommandError: Compressor is disabled. Set the COMPRESS_ENABLED setting or use --force to override.

```
```
./manage.py compress --force
Compressing... done
Compressed 23 block(s) from 122 template(s) for 1 context(s).

```

```
./manage.py collectstatic

You have requested to collect static files at the destination
location as specified in your settings:

    /encrypted/webapps/healthchecks/static-collected

This will overwrite existing files!
Are you sure you want to do this?

Type 'yes' to continue, or 'no' to cancel: yes

0 static files copied to '/encrypted/webapps/healthchecks/static-collected', 331 unmodified.

```
```
 ./manage.py migrate
Operations to perform:
  Apply all migrations: accounts, admin, api, auth, contenttypes, payments, sessions
Running migrations:
  No migrations to apply.

```
then:
`./run-gunicorn-server.sh`

which sets env and runs gunicorn - the last script used to work and it works as pings are received but web interface has HTTP 500 ERROR

**Comment by cuu508** [comment]
> [2022-11-24 12:03:56 +0000] [67790] [DEBUG] POST /ping/6cb77b0e-0006-4ef8-94d9-0ba19dd3e972

For the ping requests, were you getting HTTP status code 200 back, with the string "OK" in the response body?

**Reply by pmatuszy** [reply]
no - pings are also ending up with 500 error
```
wget https://xxxxx:8000/ping/203eadb9-300c-44bf-a3a7-b24300504fd4
--2022-11-24 14:22:44--  https://xxxxxx:8000/ping/203eadb9-300c-44bf-a3a7-b24300504fd4
Resolving xxxxx (xxxxxx)... xx.xx.xx.xx
Connecting to xxx (xxxxxxxx)|x.x.x.44|:8000... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2022-11-24 14:22:45 ERROR 500: Internal Server Error.

```

**Comment by cuu508** [comment]
What happens if you run the web server with `./manage.py runserver`? Do you also get HTTP 500 errors, anything interesting in the console?

**Reply by pmatuszy** [reply]
ok 

when I run ping from the local host:

```
wget http://127.0.0.1:8000/ping/203eadb9-300c-44bf-a3a7-b24300504fd4
--2022-11-24 14:33:54--  http://127.0.0.1:8000/ping/203eadb9-300c-44bf-a3a7-b24300504fd4
Connecting to 127.0.0.1:8000... connected.
HTTP request sent, awaiting response... 404 Not Found
2022-11-24 14:33:54 ERROR 404: Not Found.

```
I get this on the console:
```
./manage.py runserver
Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).
November 24, 2022 - 13:33:51
Django version 4.1.2, using settings 'hc.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
Not Found: /ping/203eadb9-300c-44bf-a3a7-b24300504fd4
[24/Nov/2022 13:33:54] "GET /ping/203eadb9-300c-44bf-a3a7-b24300504fd4 HTTP/1.1" 404 9

```
Though normally I use gunicorn, so I am not sure if this should work...
Paul

**Reply by cuu508** [reply]
This would be the expected response if a check with code `203eadb9-300c-44bf-a3a7-b24300504fd4` does not exist in the database.

What do you see if you run the server with `manage.py runserver` and access the web UI from the browser?

**Comment by pmatuszy** [comment]
ok - I got the loogin screen like below but my credentials which used to work don't:

![image](https://user-images.githubusercontent.com/58336955/203853076-a25878dd-24a2-418a-b10e-e09b295797bd.png)

I checked the tables in the database and I can see definitions of pings but I am not sure how shall I check logins/passwords for users...

**Comment by pmatuszy** [comment]
ok - i think there is a problem with HC version. I used to have 2.1 and it is the latest one and I can see this:

`column api_check.filter_subject does not exist LINE 1: ...grace", "api_check"."schedule", "api_check"."tz", "api_check...
`
so the database schema should be upgraded I guess. Am I right?

**Reply by cuu508** [reply]
Your computer is cursed, burn and bury it! 

But more seriously, it sounds like there are several databases, and/or several database configurations at play. Perhaps you're setting environment variables in the run-gunicorn-server script, that you are not setting otherwise. So the gunicorn process may be connecting to one database (say, postgres), while the manage.py commands connects to a different one (say, the default sqlite database)? 

You can run `manage.py dbshell` and see if the database you get connected to is the one you expect.

**Comment by pmatuszy** [comment]
@cuu508 - I fixed it. Thank you VERY MUCH for all your help and suggestions. I had to reset my mind and stay away from it for a while. Today I came back to this subject again to fix it and .... I did it.
Thank you once again for your patience.
Also you made my day with this comment: "Your computer is cursed, burn and bury it!"
:-)
Regards
Paul

**Reply by cuu508** [reply]
Happy to hear you got it resolved! What was the culprit?

**Comment by pmatuszy** [comment]
culprit? As most of the time it was "a protein interface between my ears".
I was stupid. I was not consistent with setting proper env variables. I wasted your and my time. I did it per session / by hand and when I started some new sessions or did 'su - healthcheck user' i forgot to set SOME of them. Terrible. Today I put everything in the script and .... it helped..
